proxy/http.go

```diff
@@ -0,0 +1,196 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/net/http2"
+	"html/template"
+	"net/http"
+	"net/http/httputil"
+	"strconv"
+	"strings"
+
+	"github.com/estuary/data-plane-gateway/auth"
+	"github.com/estuary/flow/go/labels"
+	log "github.com/sirupsen/logrus"
+	"github.com/valyala/fasthttp"
+	"github.com/valyala/fasthttp/fasthttpadaptor"
+)
+
+func (h *ProxyHandler) proxyHttp(ctx context.Context, clientConn *tls.Conn, proxyConn *ProxyConnection, portConfig *labels.PortConfig) error {
+	defer proxyConn.Close()
+	// We generally assume that the upstream connector container wishes to speak http/1.1, unless they explicitly request to use _only_ h2.
+	var useHttp2Upstream = portConfig != nil && portConfig.Protocol == "h2"
+	var isPublicPort = portConfig != nil && portConfig.Public
+
+	var targetScheme = "http"
+	if useHttp2Upstream {
+		targetScheme = "https"
+	}
+	var proxy = httputil.ReverseProxy{
+		Transport:     proxyConn.singleConnectionTransport(useHttp2Upstream),
+		FlushInterval: 0,
+		ErrorHandler: func(w http.ResponseWriter, req *http.Request, err error) {
+			log.WithFields(log.Fields{
+				"hostname":   proxyConn.hostname,
+				"remoteAddr": clientConn.RemoteAddr().String(),
+				"shardID":    proxyConn.shardID,
+				"error":      err,
+				"URI":        req.RequestURI,
+			}).Error("proxy error")
+			handleHttpError(err, w, req)
+		},
+
+		Director: func(req *http.Request) {
+			// TODO: verify that request Host matches the value of the connection SNI
+			req.URL.Host = proxyConn.hostname
+			req.URL.Scheme = targetScheme
+			if _, ok := req.Header["User-Agent"]; !ok {
+				// explicitly disable User-Agent so it's not set to default value
+				req.Header.Set("User-Agent", "")
+			}
+		},
+	}
+
+	var handlerFunc = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// If the port is private, then require that each request has an Authorization header that permits it to
+		// access the task. We don't check the Authorization header if the port is public, since the header value
+		// might be meant to be interpreted by the connector itself.
+		if !isPublicPort {
+			var claims, authErr = auth.AuthorizedReq(req, h.jwtVerificationKey)
+			if authErr == nil {
+				authErr = auth.EnforcePrefix(claims, proxyConn.taskName)
+			}
+			if authErr != nil {
+				handleHttpError(authErr, w, req)
+				return
+			}
+		}
+		proxy.ServeHTTP(w, req)
+	})
+
+	if clientConn.ConnectionState().NegotiatedProtocol == "h2" {
+		var h2Server = http2.Server{}
+		// The clientConn will be closed automatically by ServeConn, but we'll need to close the proxyConn ourselves
+		h2Server.ServeConn(clientConn, &http2.ServeConnOpts{
+			Context: ctx,
```

Comment:
Do you think we need any timeouts here, or just generally elsewhere? Tracing this context back its coming from a `context.Background()` higher in the stack (unless I'm missing something, which is certainly possible). I don't know that we'd be a really juicy target for a slowloris yet, but it still might be a good idea to take a principled approach time timeouts in DPG.

Reply:
I added some more specific timeouts to the HTTP servers here, which should help out. Thanks for suggesting this!